### PR TITLE
[OpenAI.Inference] Adding support for OpenAI functions.

### PR DESCRIPTION
**Pull request by glecaros**

### CODE REVIEW

File: specification/cognitiveservices/OpenAI.Inference/models/functions.tsp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using TypeSpec.Versioning;
+
+@doc("Specify the name of the only function to call.")
+@added(ServiceApiVersions.v2023_07_01_Preview)
+model FunctionName {
+  @doc("The name of the function to call.")
+  name: string;
+}
+
+@doc("Preset values to control how the mode; responds to function calls.")
```
Comment (dasMulli) [inline specification/cognitiveservices/OpenAI.Inference/models/functions.tsp:14]: typo 'mode;' -> 'model'

File: specification/cognitiveservices/OpenAI.Inference/models/functions.tsp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      """)
+  name: string;
+
+  @doc("The description of what the function does.")
+  description?: string;
+
+  #suppress "@azure-tools/typespec-azure-core/no-unknown" "External API shape takes an arbitrary json"
+  @doc("""
+      The parameters the functions accepts, described as a JSON Schema object.
+      See the [guide](/docs/guides/gpt/function-calling) for examples, and the [JSON Schema reference](https://json-schema.org/understanding-json-schema/)
```
Comment (jpalvarezl) [inline specification/cognitiveservices/OpenAI.Inference/models/functions.tsp:45]: I am not sure that the `[guide]` link will map to something valid. The prefix of that path is missing.

Comment (jpalvarezl) [reply specification/cognitiveservices/OpenAI.Inference/models/functions.tsp:45]: Here is the full link: https://platform.openai.com/docs/guides/gpt/function-calling as per the documentation: https://platform.openai.com/docs/api-reference/chat/create#chat/create-functions

File: specification/cognitiveservices/OpenAI.Inference/models/functions.tsp (hunk 1 above)
Comment (trrwilson) [reply specification/cognitiveservices/OpenAI.Inference/models/functions.tsp]: Thanks; rewrote this as:
```
@doc("""
The collection of predefined behaviors for handling request-provided function information in a chat completions
operation.
""")
```

File: specification/cognitiveservices/OpenAI.Inference/models/functions.tsp (hunk 2 above)
Comment (trrwilson) [reply specification/cognitiveservices/OpenAI.Inference/models/functions.tsp]: I opted to remove the md links and "remarks-style" references entirely for this iteration. We'll likely want to add some additional reference material a bit later on, but I suspect we'd want to link to AOAI documentation (that could in turn link to OpenAI's where appropriate). We'll need to coordinate with the right teams for that part.

**Comment by lmazuel** [comment]
We believe CI red are false positive

**Comment by scguoi** [comment]
> Hi,
> 
> I just tried to use functions with 2023-07-01-preview, and I'm still getting " Unrecognized request argument supplied: functions". Do we still need to wait for something?

Same with me. Do we still need to wait for something?

**Comment by samiur** [comment]
+1 same here.

**Comment by trrwilson** [comment]
@MysteriousAgent / @scguoi / @samiur -- thanks for asking!

Short answer: Azure OpenAI doesn't have Chat Functions turned on yet but will soon.

Longer answer: using Functions will require using the `-0613` versions of the `gpt-35-turbo` or (if separate, approved access is available) `gpt-4*` models and also needs to be enabled via a separate feature flag from the service side. Rollout on that was temporarily paused while an unrelated issue is being investigated. We don't yet have a firm ETA on when Functions will be available across all Azure OpenAI regions but I'm hopeful it'll be on the order of a few business days.

**Comment by jjczopek** [comment]
> 

If I understand correctly, separate approval is needed to get the functions available on a deployment - is that correct? once the models will be restored - what is the procedure/steps to get the functions enabled?

**Comment by jjczopek** [comment]
@trrwilson 
So if I understand correctly, when `0613` are restored and we would like to use functions API, we still need this feature to be enabled? what is the procedure/steps in such case?

**Comment by pretidav** [comment]
same question here. I am trying using api 2023-07-01-preview but I still get: 
openai.error.InvalidRequestError: Unrecognized request arguments supplied: function_call, function

Is there anything to be done to enable this functionality? Thank you

**Comment by ryriddle** [comment]
> @MysteriousAgent / @scguoi / @samiur -- thanks for asking!
> 
> Short answer: Azure OpenAI doesn't have Chat Functions turned on yet but will soon.
> 
> Longer answer: using Functions will require using the `-0613` versions of the `gpt-35-turbo` or (if separate, approved access is available) `gpt-4*` models and also needs to be enabled via a separate feature flag from the service side. Rollout on that was temporarily paused while an unrelated issue is being investigated. We don't yet have a firm ETA on when Functions will be available across all Azure OpenAI regions but I'm hopeful it'll be on the order of a few business days.

Is there an ETA for when Functions will be available now?

**Comment by YashGupta5961** [comment]
The API now supports function calling on the following models in API version 2023-07-01-preview:

- gpt-4-0613
- gpt-4-32k-0613
- gpt-35-turbo-0613
- gpt-35-turbo-16k-0613

Check this stackoverflow [post](https://stackoverflow.com/a/76563462)

**Comment by jjczopek** [comment]
> The API now supports function calling on the following models in API version 2023-07-01-preview:
> 
> 
> 
> - gpt-4-0613
> 
> - gpt-4-32k-0613
> 
> - gpt-35-turbo-0613
> 
> - gpt-35-turbo-16k-0613
> 
> 
> 
> Check this stackoverflow [post](https://stackoverflow.com/a/76563462)

Have you been able to confirm it actually works? Is it still needed to have feature enabled for function calling?

**Comment by YashGupta5961** [comment]
> 

Yep tried it with the following request:

`https://<resource>.openai.azure.com/openai/deployments/<deployment-id>/chat/completions?api-version=2023-07-01-preview`

`
{
    "messages": [{ "role": "user", "content": "What is the weather like in Boston?" }],
    "functions": [
        {
          "name": "get_current_weather",
          "description": "Get the current weather in a given location",
          "parameters": {
            "type": "object",
            "properties": {
              "location": {
                "type": "string",
                "description": "The city and state, e.g. San Francisco, CA"
              },
              "unit": {
                "type": "string",
                "enum": ["celsius", "fahrenheit"]
              }
            },
            "required": ["location"]
          }
        }
      ]
}
`

got the following response:

{
    "id": <ID>,
    "object": "chat.completion",
    "created": 1689778328,
    "model": "gpt-35-turbo-16k",
    "prompt_annotations": [
        {
            "prompt_index": 0,
            "content_filter_results": {
                "hate": {
                    "filtered": false,
                    "severity": "safe"
                },
                "self_harm": {
                    "filtered": false,
                    "severity": "safe"
                },
                "sexual": {
                    "filtered": false,
                    "severity": "safe"
                },
                "violence": {
                    "filtered": false,
                    "severity": "safe"
                }
            }
        }
    ],
    "choices": [
        {
            "index": 0,
            "finish_reason": "function_call",
            "message": {
                "role": "assistant",
                "function_call": {
                    "name": "get_current_weather",
                    "arguments": "{\n\"location\": \"Boston, MA\"\n}"
                }
            },
            "content_filter_results": {}
        }
    ],
    "usage": {
        "completion_tokens": 17,
        "prompt_tokens": 82,
        "total_tokens": 99
    }
}

**Comment by pkscanvas** [comment]
> > 
> 
> Yep tried it with the following request:
> 
> `https://<resource>.openai.azure.com/openai/deployments/<deployment-id>/chat/completions?api-version=2023-07-01-preview`
> 
> `{ "messages": [{ "role": "user", "content": "What is the weather like in Boston?" }], "functions": [ { "name": "get_current_weather", "description": "Get the current weather in a given location", "parameters": { "type": "object", "properties": { "location": { "type": "string", "description": "The city and state, e.g. San Francisco, CA" }, "unit": { "type": "string", "enum": ["celsius", "fahrenheit"] } }, "required": ["location"] } } ] }`
> 
> got the following response:
> 
> `{ "id": <ID>, "object": "chat.completion", "created": 1689778328, "model": "gpt-35-turbo-16k", "prompt_annotations": [ { "prompt_index": 0, "content_filter_results": { "hate": { "filtered": false, "severity": "safe" }, "self_harm": { "filtered": false, "severity": "safe" }, "sexual": { "filtered": false, "severity": "safe" }, "violence": { "filtered": false, "severity": "safe" } } } ], "choices": [ { "index": 0, "finish_reason": "function_call", "message": { "role": "assistant", "function_call": { "name": "get_current_weather", "arguments": "{\n\"location\": \"Boston, MA\"\n}" } }, "content_filter_results": {} } ], "usage": { "completion_tokens": 17, "prompt_tokens": 82, "total_tokens": 99 } }`

This works, just the api-version needs to be changed

**Comment by sabbadino** [comment]
It would be nice to have Ms announce support for function calling , instead of "end users" by trial and error

**Comment by ADefWebserver** [comment]
I posted a C# example at: [Recursive Azure OpenAI Function Calling](https://blazorhelpwebsite.com/ViewBlogPost/10068)

**Comment by aiakubovich** [comment]
here what is working for me:

```
openai.api_version = "2023-07-01-preview"
response = openai.ChatCompletion.create(
    engine="gpt-35-turbo-0613",
    messages=[{"role": "user", "content": "Efficient operations on the ground and friendly, attentive employees everywhere."}],
    functions=functions,
    function_call="auto",
)
```

File: specification/cognitiveservices/OpenAI.Inference/models/functions.tsp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  @doc("""
+  A description of what the function does. The model will use this description when selecting the function and
+  interpreting its parameters.
+  """)
+  description?: string;
+
+  #suppress "@azure-tools/typespec-azure-core/no-unknown" "External API shape takes an arbitrary json"
+  @doc("The parameters the functions accepts, described as a JSON Schema object.")
+  parameters?: unknown;
```
Comment (deyaaeldeen) [inline specification/cognitiveservices/OpenAI.Inference/models/functions.tsp:55]: Is there a better type to represent JSON objects? `unknown` feels unnecessarily broader type to me that may capture primitive types as well.

**Pull request merged**